lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,107 @@
+	if ( ! isset( $post->ID ) ) {
+		return;
+	}
+
+	$anchor_block_type    = 'core/navigation';
+	$first_child_position = 'first_child';
+	$last_child_position  = 'last_child';
+
+	$hooked_blocks                 = get_hooked_blocks();
+	$hooked_block_types_last_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+		? $hooked_blocks[ $anchor_block_type ][ $last_child_position ]
+		: array();
+
+	$hooked_block_types_first_child = isset( $hooked_blocks[ $anchor_block_type ][ $first_child_position ] )
+		? $hooked_blocks[ $anchor_block_type ][ $first_child_position ]
+		: array();
+
+	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, $post );
+	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, $post );
+
+	$ignored_hooked_blocks = array_merge( $hooked_block_types_last_child, $hooked_block_types_first_child );
+
+	if ( ! empty( $ignored_hooked_blocks ) ) {
+		$existing_ignored_hooked_blocks = get_post_meta( $post->ID, '_wp_ignored_hooked_blocks', true );
+		if ( ! empty( $existing_ignored_hooked_blocks ) ) {
+			$existing_ignored_hooked_blocks = json_decode( $existing_ignored_hooked_blocks, true );
+			$ignored_hooked_blocks          = array_unique( array_merge( $ignored_hooked_blocks, $existing_ignored_hooked_blocks ) );
+		}
+		update_post_meta( $post->ID, '_wp_ignored_hooked_blocks', json_encode( $ignored_hooked_blocks ) );
+	}
+}
+
+/**
+ * Hooks into the REST API response for the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_REST_Response $response The response object.
+ * @param WP_Post          $post     Post object.
+ * @param WP_REST_Request  $request  Request object.
+ * @return WP_REST_Response $response The response object.
+ */
+function gutenberg_hook_blocks_into_core_navigation_rest_response( $response, $post ) {
+	if ( ! isset( $response->data['content']['raw'] ) || ! isset( $response->data['content']['rendered'] ) ) {
+		return $response;
+	}
+	$parsed_blocks = parse_blocks( $response->data['content']['raw'] );
+	$content       = gutenberg_hook_first_last_children_into_core_navigation( $parsed_blocks, $post->ID );
+
+	// Remove mock Navigation block wrapper.
+	$start   = strpos( $content, '-->' ) + strlen( '-->' );
+	$end     = strrpos( $content, '<!--' );
+	$content = substr( $content, $start, $end - $start );
+
+	$response->data['content']['raw']      = $content;
+	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
+
+	return $response;
+}
+
+/**
+ * Adds the first and last inner blocks if there are hooked blocks.
+ *
+ * @param array $inner_blocks Parsed inner blocks of the core/navigation block.
+ * @param int $post_id Post ID of the navigation block.
+ * @return string Parsed inner blocks maybe with hooked blocks.
```

Comment:
I guess we need to update the return value description: They're no longer parsed (but serialized into a string), and we now wrap the blocks in a mock Navigation block. Maybe something like

```suggestion
 * @return string Serialized inner blocks in mock Navigation block wrapper, with hooked blocks inserted, if any.
```

Reply:
Done in https://github.com/WordPress/gutenberg/pull/57754/commits/ba5d4ca7d083c499ae195b270c81004b3147a509